refactor(core): more readability for addBatchFunc

dmaskasky opened the pull request:
## Summary
~Adds iterable to batch to make flushBatch agnostic to the internals of batch.~
edit: Just refactoring for now delaying the decision


## Check List

- [x] `pnpm run prettier` for formatting code and docs

dai-shi commented:
I thought this would decrease the bundle size, but why does this increase?

---

**Size Change:** +205 B (+0.22%) 

**Total Size:** 92.2 kB

| Filename | Size | Change |
| :--- | :---: | :---: |
| `./dist/esm/vanilla.mjs` | 4.28 kB | +40 B (+0.94%) |
| `./dist/system/vanilla.development.js` | 4.38 kB | +40 B (+0.92%) |
| `./dist/system/vanilla.production.js` | 2.13 kB | -10 B (-0.47%) |
| `./dist/umd/vanilla.development.js` | 5.72 kB | +66 B (+1.17%) |
| `./dist/umd/vanilla.production.js` | 2.86 kB | +5 B (+0.18%) |
| `./dist/vanilla.js` | 5.61 kB | +64 B (+1.15%) |

<details><summary>ℹ️ <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| `./dist/babel/plugin-debug-label.js` | 932 B |
| `./dist/babel/plugin-react-refresh.js` | 1.14 kB |
| `./dist/babel/preset.js` | 1.41 kB |
| `./dist/esm/babel/plugin-debug-label.mjs` | 1 kB |
| `./dist/esm/babel/plugin-react-refresh.mjs` | 1.19 kB |
| `./dist/esm/babel/preset.mjs` | 1.49 kB |
| `./dist/esm/index.mjs` | 62 B |
| `./dist/esm/react.mjs` | 1.4 kB |
| `./dist/esm/react/utils.mjs` | 746 B |
| `./dist/esm/utils.mjs` | 67 B |
| `./dist/esm/vanilla/utils.mjs` | 5.04 kB |
| `./dist/index.js` | 242 B |
| `./dist/react.js` | 1.44 kB |
| `./dist/react/utils.js` | 1.39 kB |
| `./dist/system/babel/plugin-debug-label.development.js` | 1.1 kB |
| `./dist/system/babel/plugin-debug-label.production.js` | 775 B |
| `./dist/system/babel/plugin-react-refresh.development.js` | 1.29 kB |
| `./dist/system/babel/plugin-react-refresh.production.js` | 928 B |
| `./dist/system/babel/preset.development.js` | 1.59 kB |
| `./dist/system/babel/preset.production.js` | 1.14 kB |
| `./dist/system/index.development.js` | 252 B |
| `./dist/system/index.production.js` | 183 B |
| `./dist/system/react.development.js` | 1.56 kB |
| `./dist/system/react.production.js` | 864 B |
| `./dist/system/react/utils.development.js` | 860 B |
| `./dist/system/react/utils.production.js` | 462 B |
| `./dist/system/utils.development.js` | 257 B |
| `./dist/system/utils.production.js` | 187 B |
| `./dist/system/vanilla/utils.development.js` | 5.25 kB |
| `./dist/system/vanilla/utils.production.js` | 3.14 kB |
| `./dist/umd/babel/plugin-debug-label.development.js` | 1.08 kB |
| `./dist/umd/babel/plugin-debug-label.production.js` | 852 B |
| `./dist/umd/babel/plugin-react-refresh.development.js` | 1.27 kB |
| `./dist/umd/babel/plugin-react-refresh.production.js` | 1 kB |
| `./dist/umd/babel/preset.development.js` | 1.54 kB |
| `./dist/umd/babel/preset.production.js` | 1.22 kB |
| `./dist/umd/index.development.js` | 383 B |
| `./dist/umd/index.production.js` | 328 B |
| `./dist/umd/react.development.js` | 1.57 kB |
| `./dist/umd/react.production.js` | 936 B |
| `./dist/umd/react/utils.development.js` | 1.53 kB |
| `./dist/umd/react/utils.production.js` | 1.01 kB |
| `./dist/umd/utils.development.js` | 399 B |
| `./dist/umd/utils.production.js` | 342 B |
| `./dist/umd/vanilla/utils.development.js` | 6.24 kB |
| `./dist/umd/vanilla/utils.production.js` | 3.78 kB |
| `./dist/utils.js` | 247 B |
| `./dist/vanilla/utils.js` | 6.1 kB |

</details>



<a href="https://github.com/preactjs/compressed-size-action"><sub>compressed-size-action</sub></a>

dai-shi commented on src/vanilla/store.ts:
```diff
@@ -645,14 +647,15 @@ const buildStore = (
           }
-        })
+        }
+        addBatchFunc(batch, processOnMount, 3)
```
Why do you prefer this to `addBatchFunc(batch, 3, processOnMount)`?

dmaskasky replied:
I think semantics, priorities could be defaulted. I think of this invocation as _Add function to batch [with priority]_.
I would swap position of batch and func, but I think of batch as a form of context and it makes ordering easier for more complex function signatures.

dai-shi replied:
I would like to prohibit defaulting priority, and `addBatchFunc(batch, priority, fn)` feels more functional style.

dmaskasky replied:
Ok, I'll change it back next year.

dmaskasky commented:
> I thought this would decrease the bundle size, but why does this increase?

- I had added a placeholder for syncEffect, but I realized we should do that in the syncEffect PR, so I reverted it just now
- replacing inlined functions with name functions
- `Omit<ReadonlyArray<Set<() => void>>, number> & { length: 3 }` is to preserve jsdoc comments while retaining tuple-like functionality

dai-shi commented on src/vanilla/store.ts:
```diff
@@ -163,30 +163,24 @@ const addDependency = <Value>(
 //
 
-type BatchPriority = 'H' | 'M' | 'L'
+type BatchPriority = 0 | 1 | 2
```
I think we can use numbers as object keys (internally they will be converted into strings.)

So, my proposition is:
```ts
type Batch = Readonly<{
  0: Set<() => void>
  1: Set<() => void>
  2: Set<() => void>
  D: Map<AnyAtom, Set<AnyAtom>>
}>
```

dai-shi replied:
https://github.com/pmndrs/jotai/pull/2898/files#r1900329774

If we were to generalize the batch queues, I may consider a single ordered queue. Not sure about the performance or the bungle size, though.

dmaskasky replied:
A queue queue might be nice. We wouldn't need explicit placeholders defined in the store for third-party applications.
Its a little more on bundle size though.

Rough:
```ts
const batch = (() => {
  const H = new Set()
  const M = new Set()
  const L = new Set()
  const queue = [H, M, L]
  const D = new Map()
  const batch = Object.assign(queue, { D, H, M, L })
  return batch
})()
...
const syncEffect = new Set()
// schedule sync effect after H, but before M
const hIndex = batch.findIndex(element => element === batch.H)
batch.splice(hIndex + 1, 0, syncEffect)
```

dai-shi replied:
I meant a single queue. But don't take it seriously as it may have performance drawbacks.

dai-shi replied:
For now, I think queue array is better. 8945774

dmaskasky replied:
Yes that is what I am proposing, but your implementation does not allow for "insert queue after..." capability.

dai-shi replied:
In that case, I'd consider a single priority queue.

dai-shi commented on src/vanilla/store.ts:
```diff
@@ -231,14 +226,12 @@ const flushBatch = (batch: Batch) => {
-    batch.M.clear()
-    batch.L.forEach(call)
-    batch.L.clear()
+    for (const channel of batch) {
```
Oh, I see. You want to iterate.

dai-shi commented:
https://github.com/pmndrs/jotai/actions/runs/12575899712/job/35051418426?pr=2898

**Size Change:** +175 B (+0.19%) 

**Total Size:** 92.2 kB

| Filename | Size | Change |
| :--- | :---: | :---: |
| `./dist/esm/vanilla.mjs` | 4.28 kB | +35 B (+0.82%) |
| `./dist/system/vanilla.development.js` | 4.37 kB | +31 B (+0.71%) |
| `./dist/system/vanilla.production.js` | 2.15 kB | +1 B (+0.05%) |
| `./dist/umd/vanilla.development.js` | 5.71 kB | +55 B (+0.97%) |
| `./dist/umd/vanilla.production.js` | 2.85 kB | +3 B (+0.11%) |
| `./dist/vanilla.js` | 5.6 kB | +50 B (+0.9%) |

<details><summary>ℹ️ <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| `./dist/babel/plugin-debug-label.js` | 932 B |
| `./dist/babel/plugin-react-refresh.js` | 1.14 kB |
| `./dist/babel/preset.js` | 1.41 kB |
| `./dist/esm/babel/plugin-debug-label.mjs` | 1 kB |
| `./dist/esm/babel/plugin-react-refresh.mjs` | 1.19 kB |
| `./dist/esm/babel/preset.mjs` | 1.49 kB |
| `./dist/esm/index.mjs` | 62 B |
| `./dist/esm/react.mjs` | 1.4 kB |
| `./dist/esm/react/utils.mjs` | 746 B |
| `./dist/esm/utils.mjs` | 67 B |
| `./dist/esm/vanilla/utils.mjs` | 5.04 kB |
| `./dist/index.js` | 242 B |
| `./dist/react.js` | 1.44 kB |
| `./dist/react/utils.js` | 1.39 kB |
| `./dist/system/babel/plugin-debug-label.development.js` | 1.1 kB |
| `./dist/system/babel/plugin-debug-label.production.js` | 775 B |
| `./dist/system/babel/plugin-react-refresh.development.js` | 1.29 kB |
| `./dist/system/babel/plugin-react-refresh.production.js` | 928 B |
| `./dist/system/babel/preset.development.js` | 1.59 kB |
| `./dist/system/babel/preset.production.js` | 1.14 kB |
| `./dist/system/index.development.js` | 252 B |
| `./dist/system/index.production.js` | 183 B |
| `./dist/system/react.development.js` | 1.56 kB |
| `./dist/system/react.production.js` | 864 B |
| `./dist/system/react/utils.development.js` | 860 B |
| `./dist/system/react/utils.production.js` | 462 B |
| `./dist/system/utils.development.js` | 257 B |
| `./dist/system/utils.production.js` | 187 B |
| `./dist/system/vanilla/utils.development.js` | 5.25 kB |
| `./dist/system/vanilla/utils.production.js` | 3.14 kB |
| `./dist/umd/babel/plugin-debug-label.development.js` | 1.08 kB |
| `./dist/umd/babel/plugin-debug-label.production.js` | 852 B |
| `./dist/umd/babel/plugin-react-refresh.development.js` | 1.27 kB |
| `./dist/umd/babel/plugin-react-refresh.production.js` | 1 kB |
| `./dist/umd/babel/preset.development.js` | 1.54 kB |
| `./dist/umd/babel/preset.production.js` | 1.22 kB |
| `./dist/umd/index.development.js` | 383 B |
| `./dist/umd/index.production.js` | 328 B |
| `./dist/umd/react.development.js` | 1.57 kB |
| `./dist/umd/react.production.js` | 936 B |
| `./dist/umd/react/utils.development.js` | 1.53 kB |
| `./dist/umd/react/utils.production.js` | 1.01 kB |
| `./dist/umd/utils.development.js` | 399 B |
| `./dist/umd/utils.production.js` | 342 B |
| `./dist/umd/vanilla/utils.development.js` | 6.24 kB |
| `./dist/umd/vanilla/utils.production.js` | 3.78 kB |
| `./dist/utils.js` | 247 B |
| `./dist/vanilla/utils.js` | 6.1 kB |

</details>



<a href="https://github.com/preactjs/compressed-size-action"><sub>compressed-size-action</sub></a>

---

The increase may also come from explicit naming callback functions.

dai-shi commented:
Using iterable batch is still under consideration. However, I'm leaning towards it.
As it's very related with onInit and sync effect, I'd like to delay the decision.
Meanwhile, I want to release a patch. So, let me tweak this PR for a simple refactor. (and acceptable bundle increases.)

dai-shi commented on src/vanilla/store.ts:
```diff
@@ -165,25 +165,38 @@ const addDependency = <Value>(
-  M: new Set(),
-  L: new Set(),
-})
+const createBatch = (): Batch => {
+  const batch = { D: new Map() } as {
+    D: Map<AnyAtom, Set<AnyAtom>>
+    [BATCH_PRIORITY_HIGH]: Set<() => void>
+    [BATCH_PRIORITY_MEDIUM]: Set<() => void>
+    [BATCH_PRIORITY_LOW]: Set<() => void>
+  }
+  batch[BATCH_PRIORITY_HIGH] = new Set()
+  batch[BATCH_PRIORITY_MEDIUM] = new Set()
+  batch[BATCH_PRIORITY_LOW] = new Set()
+  return batch
+}
```
I'm not super confident with it, but this seems better for minification.
<img width="1316" alt="image" src="https://github.com/user-attachments/assets/3496ca35-9fb4-47bf-831a-4e9961060aa7" />

We will be revisiting it soon with sync effect anyway, otherwise, we can revert it back.
